finagle/buoyant/src/main/scala/com/twitter/finagle/buoyant/PathMatcher.scala

```diff
@@ -58,33 +58,47 @@ object PathMatcher {
 
   private class Matcher(expr: String) extends PathMatcher {
 
-    val exprSegments = expr.split("/").dropWhile(_.isEmpty)
-
     override def extract(path: Path): Option[Map[String, String]] =
-      _extract(path.showElems, exprSegments, Map.empty)
+      _extract(path.showElems, regexSegments, Map.empty)
+
+    private[this] val segmentRegex = """\{([a-zA-Z0-9\.:-]+)\}""".r
+
+    private[this] def mkRegex(x: String, y: String) = x.replace(y, "(.*)")
+
+    private[this] val regexSegments: Seq[(Option[scala.util.matching.Regex], Seq[String])] = {
+      expr.split("/").dropWhile(_.isEmpty).map { exprSegment =>
+        segmentRegex.findAllIn(exprSegment) match {
+          case k if !k.isEmpty =>
+            var keys = k.toArray
+            Some(keys.foldLeft(exprSegment.replace(".", "\\."))(mkRegex).r) -> keys.map { key =>
```

Comment:
there's a lot going on in this line which makes it hard to understand.  consider breaking it up:

```
// '.' is the only valid Path character that also has a regex meaning.  Any '.' characters must be escaped.
val escaped = exprSegment.replace(".", """\.""") 
// Replace capture keys with regex captures.
val withCaptures = keys.foldLeft(escaped)(_.replace(_, "(.*)")
val keyNames = keys.map { key =>
  key.drop(1).dropRight(1) // Remove the encasing '{' and '}' characters.
}.toSeq
```